File: leetcode_39/combinationSum.go
package main

import "sort"

func getRes(arr []int, target int, total int, start int, path []int, res *[][]int) {
    if total > target {
        return
    }
    if total == target {
        t := make([]int, len(path))
        copy(t, path)
        *res = append(*res, t)
        return
    }

    for i := start; i < len(arr); i++ {
        if total + arr[i] > target {
            continue
        }
        path = append(path, arr[i])
        getRes(arr, target, total + arr[i], i, path, res)
        path = path[ : len(path) - 1]
    }
}

// combinationSum: 重点是先进行一次排序，并且不回头查看
func combinationSum(candidates []int, target int) [][]int {
    res := make([][]int, 0)
    sort.Ints(candidates)
    getRes(candidates, target, 0, 0, []int{}, &res)
    return res
}

File: leetcode_39/CombinationSum.java
package leetcode_39;

import java.util.*;

public class CombinationSum {

    private Node root;
    private List<Integer> combines = new LinkedList<>();
    private List<List<Integer>> result = new ArrayList<>();

    private class Node {
        int total = 0;
        Node[] children;
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        root = getCombine(root, candidates, target, 0, 0);
        getResult(root, candidates, target);
        return result;
    }

    public Node getCombine(Node x, int[] candidates, int target, int curt, int start) {
        int N = candidates.length;
        x = new Node();
        x.total = curt;
        x.children = new Node[N];
        for(int i = start; i < N; i++) {
            if(curt + candidates[i] <= target) {
                x.children[i] = getCombine(x.children[i], candidates, target, curt + candidates[i], i);
            } else {
                break;
            }
        }
        return x;
    }

    public void getResult(Node x, int[] candidates, int target) {
        if(x == null) {
            return;
        }
        if(isGood(x, target)) {
            result.add(List.copyOf(combines));
            for(Integer t: combines) {
                System.out.print(t + " ");
            }
            System.out.println();
        }
        for(int i = 0; x.children != null && i < x.children.length; i++) {
            combines.add(candidates[i]);
            getResult(x.children[i], candidates, target);
            combines.remove(combines.size() - 1);
        }
    }

    private boolean isGood(Node x, int target) {
        return x.total == target;
    }

    public static void main(String[] args) {
        int[] test = new int[]{10, 1, 2, 7, 6, 1, 5};
        CombinationSum com = new CombinationSum();
        com.combinationSum(test, 8);
    }
}
